ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Transaction.java

```diff
@@ -634,32 +637,44 @@ private static Bytes frontierPreimage(
 
   private static Bytes eip1559Preimage(
       final long nonce,
-      final Wei gasPrice,
       final Wei gasPremium,
       final Wei feeCap,
       final long gasLimit,
       final Optional<Address> to,
       final Wei value,
       final Bytes payload,
-      final Optional<BigInteger> chainId) {
-    return RLP.encode(
-        rlpOutput -> {
-          rlpOutput.startList();
-          rlpOutput.writeLongScalar(nonce);
-          rlpOutput.writeUInt256Scalar(gasPrice);
-          rlpOutput.writeLongScalar(gasLimit);
-          rlpOutput.writeBytes(to.map(Bytes::copy).orElse(Bytes.EMPTY));
-          rlpOutput.writeUInt256Scalar(value);
-          rlpOutput.writeBytes(payload);
-          rlpOutput.writeUInt256Scalar(gasPremium);
-          rlpOutput.writeUInt256Scalar(feeCap);
-          if (chainId.isPresent()) {
-            rlpOutput.writeBigIntegerScalar(chainId.get());
-            rlpOutput.writeUInt256Scalar(UInt256.ZERO);
-            rlpOutput.writeUInt256Scalar(UInt256.ZERO);
-          }
-          rlpOutput.endList();
-        });
+      final Optional<BigInteger> chainId,
+      final Optional<List<AccessListEntry>> accessList) {
+    final Bytes encoded =
+        RLP.encode(
+            rlpOutput -> {
+              rlpOutput.startList();
+              rlpOutput.writeBigIntegerScalar(chainId.orElseThrow());
+              rlpOutput.writeLongScalar(nonce);
+              rlpOutput.writeUInt256Scalar(gasPremium);
+              rlpOutput.writeUInt256Scalar(feeCap);
+              rlpOutput.writeLongScalar(gasLimit);
+              rlpOutput.writeBytes(to.map(Bytes::copy).orElse(Bytes.EMPTY));
+              rlpOutput.writeUInt256Scalar(value);
+              rlpOutput.writeBytes(payload);
+              if (accessList.isPresent()) {
+                rlpOutput.writeList(
+                    accessList.get(),
+                    (accessListEntry, accessListEntryRLPOutput) -> {
+                      accessListEntryRLPOutput.startList();
+                      rlpOutput.writeBytes(accessListEntry.getAddress());
+                      rlpOutput.writeList(
+                          accessListEntry.getStorageKeys(),
+                          (storageKeyBytes, storageKeyBytesRLPOutput) ->
+                              storageKeyBytesRLPOutput.writeBytes(storageKeyBytes));
+                      accessListEntryRLPOutput.endList();
```

Comment:
We should probably deduplicate this between here and `TransactionEncoder::encodeAccessListInner`

Reply:
Yes

Reply:
Done https://github.com/hyperledger/besu/pull/2043/commits/093e27344a6aa281edb14edc75169c73bf08c910

Reply:
Can we not use the method you introduced here if we use `accessList.ifPresent`?

Reply:
Done